From 3cb8e35f6adba3b0ead0ac2d1e8327389227cd47 Mon Sep 17 00:00:00 2001
From: Daniel Borkmann <dborkman@redhat.com>
Date: Wed, 18 Dec 2013 13:53:02 -0500
Subject: [tools] bpf_dbg: always close socket in bpf_runnable

Message-id: <1758f642f6701451e5d571d41c1061ca3686d124.1387364406.git.dborkman@redhat.com>
Patchwork-id: 72493
O-Subject: [RHEL7 PATCH net 4/5] bpf_dbg: always close socket in bpf_runnable
Bugzilla: 958035
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Thomas Graf <tgraf@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=958035
Upstream Status: 8b138da62f3ef370bebaaed7385020993bb787ae (net-next)
Brew: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=6751484
Tested: verified on beaker

Upstream Commit:
commit 8b138da62f3ef370bebaaed7385020993bb787ae
Author: Daniel Borkmann <dborkman@redhat.com>
Date:   Mon Dec 16 11:45:00 2013 +0100

    bpf_dbg: always close socket in bpf_runnable

    We must not leave the socket intact in bpf_runnable(). The socket
    is used to test if the filter code is being accepted by the kernel
    or not. So right after we do the setsockopt(2), we need to close
    it again.

    Signed-off-by: Daniel Borkmann <dborkman@redhat.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Daniel Borkmann <dborkman@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/net/bpf_dbg.c b/tools/net/bpf_dbg.c
index 0fdcb70..65dc757 100644
--- a/tools/net/bpf_dbg.c
+++ b/tools/net/bpf_dbg.c
@@ -512,11 +512,11 @@ static bool bpf_runnable(struct sock_filter *f, unsigned int len)
   return false;
  }
  ret = setsockopt(sock, SOL_SOCKET, SO_ATTACH_FILTER, &bpf, sizeof(bpf));
+ close(sock);
  if (ret < 0) {
   rl_printf("program not allowed to run by kernel!\n");
   return false;
  }
- close(sock);
  for (i = 0; i < len; i++) {
   if (BPF_CLASS(f[i].code) == BPF_LD &&
       f[i].k > SKF_AD_OFF) {
-- 
1.7.1